<a href="https://colab.research.google.com/github/criszhirzhan/maestria_PLN_resources/blob/main/caracteristicasOdio_26.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural**

***

---

Datos del alumno: Christian Zhirzhan

Fecha:
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Caracterización de textos</span>

**Objetivos**

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Primero se debe desacargar el modelo ya que no vienen por defecto en spacy.

In [1]:
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 21.1 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


Observa las diferentes librerías que se están importando.

In [2]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>:

https://spacy.io/models/es#es_core_news_md


In [3]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [4]:
filename = "/content/02Dataset_anonimizado.csv"
lines_number = 10000
# Cargar TODO el dataset
data_full = pd.read_csv(filename, delimiter=';', encoding='latin-1')

# Tomar muestra aleatoria de 10,000 registros
data = data_full.sample(n=lines_number, random_state=42)


print("Tamaño del dataset completo:", data_full.shape)
print("Tamaño de la muestra:", data.shape)

data.head()

#Se validan que existan las columnas:

print(data.columns)

/tmp/ipykernel_5752/2564266614.py:4: DtypeWarning: Columns (6,7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  data_full = pd.read_csv(filename, delimiter=';', encoding='latin-1')


Tamaño del dataset completo: (575029, 16)
Tamaño de la muestra: (10000, 16)
Index(['MEDIO', 'SOPORTE', 'URL', 'TIPO DE MENSAJE', 'CONTENIDO A ANALIZAR',
       'INTENSIDAD', 'TIPO DE ODIO', 'TONO HUMORISTICO', 'MODIFICADOR',
       'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12',
       'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15'],
      dtype='object')


El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas:


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos

In [ ]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):

    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)

    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra
for token in doc[1]:
    print(token)

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [5]:
print("Número de registros en la muestra:", len(data))
print("Número de registros total:", len(data_full))

Número de registros en la muestra: 10000
Número de registros total: 575029


<b>Se usa el atributo shape del DataFrame porque este devuelve el total de filas y columnas, por lo que en el resultado se puede visualizar el total de registros en filas (filas, columnas) dando como resultado 10000 para la muestra y 575029 para los registros totales.</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [7]:
###################### EJEMPLO PARA CONTEO ############

texto_prueba = "Hola, mundo!? Esto es una prueba."

doc = nlp(texto_prueba)

tokens_filtrados = [
    token.text for token in doc
    if not token.is_punct and not token.is_space
]

print("EJEMPLO DE CONTEO")
print("Texto original:", texto_prueba)
print("Tokens totales:", [token.text for token in doc])
print("Tokens filtrados (solo palabras):", tokens_filtrados)
print("Número de palabras:", len(tokens_filtrados))

######################## CONTEO DEL CORPUS ################################
print("------------------------------------------------------")

total_palabras = data["CONTENIDO A ANALIZAR"].apply(
    lambda x: len([
        token for token in nlp(str(x))
        if not token.is_punct and not token.is_space
    ])
).sum()
print("CONTEO DEL CORPUS")
print("Total de palabras:", total_palabras)

EJEMPLO DE CONTEO
Texto original: Hola, mundo!? Esto es una prueba.
Tokens totales: ['Hola', ',', 'mundo', '!', '?', 'Esto', 'es', 'una', 'prueba', '.']
Tokens filtrados (solo palabras): ['Hola', 'mundo', 'Esto', 'es', 'una', 'prueba']
Número de palabras: 6
------------------------------------------------------
CONTEO DEL CORPUS
Total de palabras: 1144261


<b>Para realizar el conteo de palabras reales, excluyendo los signos de puntuación y los espacios, spaCy no lo hace por sí solo. Para ello se pueden usar los atributos is_punct e is_space, que permiten excluir signos y espacios y así obtener un conteo de palabras más preciso, esto da como resultado el valor de 1144261 palabras para el corpus de 10k datos, tambien se agrego un ejemplo para validar como funciona el conteo de palabras con spacy usando sus atributos. </b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [ ]:
data["num_palabras"] = data["CONTENIDO A ANALIZAR"].apply(
    lambda x: len([
        token for token in nlp(str(x))
        if not token.is_punct and not token.is_space
    ])
)

promedio = data["num_palabras"].mean()

print("Promedio de palabras:", promedio)

Promedio de palabras: 114.4261


<b>Para calcular el promedio de palabras en cada comentario primero se eliminan los signos de puntuación y los espacios utilizando las herramientas que ofrece spaCy como is_punct y is_space, después se cuentan las palabras de cada comentario y ese número se guarda en una nueva columna del DataFrame llamada “num_palabras”, con esta columna se pueden realizar distintos cálculos, como obtener el promedio de palabras, que en este caso indica cuántas palabras tiene en promedio cada oración con la funcion mean(). </b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [ ]:

# Se crea la columna odio en el DataFrame la cual guarda valores binarios.
  # 1 -> Cuando el valor de intensidad es mayor a cero (Odio)
  # 0 -> Cuando el valor de intensidad es cero (No Odio).

data["ODIO"] = data["INTENSIDAD"].apply(lambda x: 0 if x == 0 else 1)


promedio_no_odio = data[data["ODIO"] == 0]["num_palabras"].mean()
promedio_odio = data[data["ODIO"] == 1]["num_palabras"].mean()

print("Promedio palabras (No odio):", promedio_no_odio)
print("Promedio palabras (Odio):", promedio_odio)

Promedio palabras (No odio): 115.52089506488198
Promedio palabras (Odio): 64.12206572769954


<b>Para clasificar el promedio de palabras de los comentarios en los grupos de odio y no odio, primero se crea una nueva columna llamada “ODIO” en el DataFrame. Esta columna almacena valores binarios según la intensidad: se asigna 0 cuando la intensidad es igual a cero (no odio) y 1 cuando la intensidad es mayor que cero (odio). Luego, se utiliza la columna “num_palabras”, creada anteriormente, que contiene el número de palabras de cada comentario, para calcular el promedio de palabras en cada uno de estos grupos.</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>


<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Plantea tus propias preguntas</span>

<span><b>Plantea al menos 4 características</b> del texto cuyo análisis permita una caracterización completa del texto. Puedes utilizar recomendaciones proporcionadas por la IA Generativa, si así lo deseas. Para cada una de las características planteadas, obtén valores separados para los grupos ODIO/NO-ODIO.</span>

<span>En la explicación aportada, deberás <b>explicar el significado de la característica planteada</b> así como la importancia de ésta en la caracterización del texto.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 1.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[INDICA AQUÍ LA CARACTERÍSTICA QUE PROPONES]</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 2.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[INDICA AQUÍ LA CARACTERÍSTICA QUE PROPONES]</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 3.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[INDICA AQUÍ LA CARACTERÍSTICA QUE PROPONES]</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 4.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[INDICA AQUÍ LA CARACTERÍSTICA QUE PROPONES]</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr>


<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Reflexión final</span>

<span>Una de las utilidades de la caracterización de texto es que nos sirve como etapa de <i>feature-extraction</i> (extración de características) de cara a un posterior sistema de clasificación. Es pertinente, por tanto, reflexionar sobre la capacidad discriminatoria de cada una de las características extraídas. </span>

<span> Responde, para ello, a la siguiente pregunta.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Reflexión final.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [ ]:
# Incluye aquí el código generado para poder responder a tu pregunta


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
